pilot/pkg/config/clusterregistry/secretcontroller.go

```diff
@@ -167,119 +165,111 @@ func (c *Controller) runWorker() {
 }
 
 func (c *Controller) processNextItem() bool {
-	key, quit := c.queue.Get()
+	secretName, quit := c.queue.Get()
 
 	if quit {
 		return false
 	}
-	defer c.queue.Done(key)
+	defer c.queue.Done(secretName)
 
-	err := c.processItem(key.(string))
+	err := c.processItem(secretName.(string))
 	if err == nil {
 		// No error, reset the ratelimit counters
-		c.queue.Forget(key)
-	} else if c.queue.NumRequeues(key) < maxRetries {
-		log.Errorf("Error processing %s (will retry): %v", key, err)
-		c.queue.AddRateLimited(key)
+		c.queue.Forget(secretName)
+	} else if c.queue.NumRequeues(secretName) < maxRetries {
+		log.Errorf("Error processing %s (will retry): %v", secretName, err)
+		c.queue.AddRateLimited(secretName)
 	} else {
-		log.Errorf("Error processing %s (giving up): %v", key, err)
-		c.queue.Forget(key)
+		log.Errorf("Error processing %s (giving up): %v", secretName, err)
+		c.queue.Forget(secretName)
 		utilruntime.HandleError(err)
 	}
 
 	return true
 }
 
-func (c *Controller) processItem(key string) error {
-	obj, exists, err := c.informer.GetIndexer().GetByKey(key)
+func (c *Controller) processItem(secretName string) error {
+	obj, exists, err := c.informer.GetIndexer().GetByKey(secretName)
 	if err != nil {
-		return fmt.Errorf("error fetching object with key %s from store: %v", key, err)
+		return fmt.Errorf("error fetching object %s from store: %v", secretName, err)
 	}
 
 	if !exists {
-		c.secretDelete(key)
+		c.deleteMemberCluster(secretName)
 		return nil
 	}
-	c.secretAdd(obj)
+	c.addMemberCluster(secretName, obj.(*corev1.Secret))
 
 	return nil
 }
 
-func addMemberCluster(s *corev1.Secret, c *Controller) {
+func (c *Controller) addMemberCluster(secretName string, s *corev1.Secret) {
 	c.cs.storeLock.Lock()
 	defer c.cs.storeLock.Unlock()
 	// Check if there is already a cluster member with the specified
-	key := Metadata{Name: s.ObjectMeta.Name, Namespace: s.ObjectMeta.Namespace}
```

Comment:
@baodongli how about keep this logic, I think your PR will be smaller without this change.

Based on comments from @sbezverk , I think it is better not to touch this logic to avoid future refactor.

Reply:
it doesn't make sense to put a namespace in front of something that's not even a k8s object to make it unique. One would be puzzled while reading the code. 

Right now, it doesn't even watch secrets from more than one namespace. So why the namespace is used as part of the key in the first place.

The secret with the muticluster label is just a way to carry kubeconfig files that configure access to remote clusters, the secret itself is not equivalent to a cluster object. You can create two secrets in the same namespace that stores the same kubeconfig files. Multiple kubeconfig files can be included in the same secret. Uniqueness of kubeconfig files can't be guaranteed by the way they are configured now.

Reply:
by design there was supposed to be 1 secret objet for 1 cluster, having multiple kubeconfigs in a single secret was never discussed and of course the controller was not programmed for this.

Reply:
@sbezverk @baodongli For the namespace issue, considering it does not impact much, so how about we merge this PR first and open an issue to have more discussion there? The pilot crash is a serious issue and I hope that we can fix this asap.

/cc @sdake